### DISCUSSION TITLE
Visidata plugin for Neovim

### DISCUSSION THREAD

**Discussion by CharlesARoy** [saulpw/visidata#1511] [Ideas]
Visidata and Neovim are both amazing tools for working on the command line. I'd love to see a Visidata plugin for Neovim so that it can be used directly inside Neovim. (In the meantime, I can keep loading it with [Toggleterm](https://github.com/akinsho/toggleterm.nvim)).

**Comment by abitrolly** [comment]
Do you mean opening `vd` when the file type is `.csv`? Or how it should work? Can you clarify the workflow from the moment of plugin installation?

**Reply by CharlesARoy** [reply]
I'm sure that the pairing would eventually be much more powerful than this simple use case, but this is what I had in mind:

Currently, if I open a tabular file in Neovim, it might look something like this:
![image](https://user-images.githubusercontent.com/46253001/189184959-1f8220b5-81dc-43e3-ac60-179f2396bfd2.png)

If I want, I can use the Toggleterm plugin to launch a terminal inside Neovim and then load the file with Visidata like this:
![image](https://user-images.githubusercontent.com/46253001/189185331-01e16f3d-f537-44b8-8962-f24e22b026bc.png)

However, it would be nice if, after opening the file, I could set some keymaps to load it in Visidata directly in a Neovim buffer. 

The workflow might look something like this:
1. I have several files open in Neovim and am adverse to leaving it.
2. I open a new file and see that it's tabular data (and is poorly formatted in Neovim).
3. I type some key combo, e.g. `<Leader>v` and the buffer is converted to a Visidata page showing the same file.
4. I continue working on my other Neovim files, easily switching back and forth to the buffer in the Visidata view.

**Reply by abitrolly** [reply]
I guess creating the keymap to pass the buffer to `vd` via Toggleterm is not a problem. Is the problem that Toggleterm opens a new system window and now you need to switch between those using non-vim keys?

**Reply by rishabhjain4** [reply]
Instead of a plugin, It's better to use nvim with tmux. You can always have a new screen for visidata to work on. Also, keep the zoom keymap for tmux screen handy for this. I have been using this way.

**Reply by gennaro-tedesco** [reply]
> I type some key combo, e.g. <Leader>v and the buffer is converted to a Visidata page showing the same file.

what do you mean by "buffer being converted to a Visidata page"? Visidata being a terminal application, it needs to live in a terminal, so at most you can have it in a neovim _terminal buffer_. In which case why is the solution of having a keymap opening a terminal buffer and loading the sheet in Visidata not what you want?

**Reply by CharlesARoy** [reply]
@gennaro-tedesco your solution is likely sufficient and the entire request was probably misguided. These days I mostly load Nvim and Visidata in separate terminal windows or use [fancy-ctrl-z](https://github.com/ohmyzsh/ohmyzsh/tree/master/plugins/fancy-ctrl-z) to send Nvim to the background, do what I need to do in Visidata, and then bring it back to the foreground.

**Comment by emilBeBri** [comment]
I'm struggling with the similar problem, and as you seem to be on yourself, @CharlesARoy , it's likely a problem of not using the right concepts for the right job, so to speak. If you have the time, would you mind showing your solution/workflow to achieve this? It does not seem overly complicated but I'm struggling with it conceptually, so a practical solution would help.

**Comment by asteroidb612** [comment]
Here's a line from my init.vim for using visidata within nvim. 

```
"I want my yank / paste to use the system copy/paste clipboard
set clipboard=unnamed

" I want to open the contents of this buffer in vd
nnoremap <Leader>v ggyG<C-o>:tabnew<CR>:terminal<CR>apbpaste \| vd<CR>
```

This fits my workflow, where I use nvim's tabs as my window manager (rather than tmux). 

Note, this is on MacOS (that pbpaste above is mac specific, I believe), and may not be helpful outside that.

**Comment by mairs8** [comment]
https://github.com/Willem-J-an/visidata.nvim
